File: cs/clean-code/Chess/Board.cs
using System.Collections.Generic;
using System.Linq;

namespace Chess
{
    public class Board
    {
        private readonly Piece[][] cells;

        public Board(Piece[][] cells)
        {
            this.cells = cells;
        }

        public IEnumerable<Location> GetPieces(PieceColor color) => 
            AllBoard().Where(loc => Piece.Is(GetPiece(loc), color));

        public Piece GetPiece(Location location) => 
            Contains(location) ? cells[location.Y][location.X] : null;

        public void Set(Location location, Piece cell) => 
            cells[location.Y][location.X] = cell;

        public TemporaryPieceMove PerformTemporaryMove(Location from, Location to)
        {
            var old = GetPiece(to);
            Set(to, GetPiece(from));
            Set(from, null);
            return new TemporaryPieceMove(this, from, to, old);
        }

        private IEnumerable<Location> AllBoard()
        {
            for (int y = 0; y < cells.Length; y++)
            for (int x = 0; x < cells[0].Length; x++)
                yield return new Location(x, y);
        }

        public bool Contains(Location loc) =>
            loc.X >= 0 && loc.X < cells[0].Length && 
            loc.Y >= 0 && loc.Y < cells.Length;
    }
}

File: java/src/main/java/ru/kontur/courses/chess/Board.java
package ru.kontur.courses.chess;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Board {
    private final Piece[][] cells;

    public Board(Piece[][] cells) {
        this.cells = cells;
    }

    public Iterable<Location> getPieces(PieceColor color) {
        return allBoards().filter(it -> Piece.is(getPiece(it), color)).collect(Collectors.toList());
    }

    public Piece getPiece(Location location) {
        return contains(location) ? cells[location.y()][location.x()] : null;
    }

    public void set(Location location, Piece cell) {
        cells[location.y()][location.x()] = cell;
    }

    public TemporaryPieceMove performTemporaryMove(Location from, Location to) {
        var old = getPiece(to);
        set(to, getPiece(from));
        set(from, null);
        return new TemporaryPieceMove(this, from, to, old);
    }

    public Stream<Location> allBoards() {
        var result = new ArrayList<Location>();
        for (int y = 0; y < cells.length; y++)
            for (int x = 0; x < cells[0].length; x++) {
                result.add(new Location(x, y));
            }

        return result.stream();
    }

    public boolean contains(Location loc) {
        return loc.x() >= 0 && loc.x() < cells[0].length &&
                loc.y() >= 0 && loc.y() < cells.length;
    }
}
